source/Font.cpp

```diff
@@ -12,91 +12,169 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
-		"in vec2 texCoord;\n"
+		// invalid - 111????? 10?????? invalid
+		if((*(str + 2) & 0xc0) != 0x80)
+			return -1;
 		
-		// Output color.
-		"out vec4 finalColor;\n"
+		// 3 bytes - 1110xxxx 10xxxxxx 10xxxxxx
+		if((*str & 0x10) == 0)
+			return 3;
 		
-		// Multiply the texture by the user-specified color (including alpha).
-		"void main() {\n"
-		"  finalColor = texture(tex, texCoord).a * color;\n"
-		"}\n";
-	
-	const int KERN = 2;
+		// invalid - 1111???? 10?????? 10?????? invalid
+		if((*(str + 3) & 0xc0) != 0x80)
+			return -1;
+		
+		// 4 bytes - 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+		if((*str & 0x8) == 0)
+			return 4;
+		
+		// not unicode - 11111??? 10?????? 10?????? 10??????
+		return -1;
+	}
 }
 
 
 
 Font::Font()
-	: texture(0), vao(0), vbo(0), colorI(0), scaleI(0), glyphI(0), aspectI(0),
-	  positionI(0), height(0), space(0), screenWidth(0), screenHeight(0)
+	: size(0)
 {
 }
 
 
 
-Font::Font(const string &imagePath)
-	: Font()
+bool Font::Load(const DataNode &node)
 {
-	Load(imagePath);
+	// Ignore if already loaded.
+	if(!sources.empty())
+		return false;
+	
+	if(node.Size() < 1 || node.Token(0) != "font")
+	{
+		node.PrintTrace("was expecting font");
+		return false;
+	}
+	
+	// Get size.
+	size = 0;
+	for(const DataNode &child : node)
+	{
+		string key = (child.Size() >= 0 ? child.Token(0) : "");
+		if(key != "size")
+			continue;
+		
+		if(size > 0)
+		{
+			node.PrintTrace("too many sizes");
+			return false;
+		}
+		size = round(child.Value(1));
+		if(size <= 0)
+		{
+			child.PrintTrace("invalid font.size");
+			return false;
+		}
+	}
+	if(size <= 0)
+	{
+		node.PrintTrace("missing size");
+		return false;
+	}
+	
+	// Get glyph source.
+	sources.clear();
+	for(const DataNode &child : node)
+	{
+		if(child.Size() <= 0 || (child.Token(0) != "atlas" && child.Token(0) != "freetype"))
+			continue;
+		string key = child.Token(0);
+		
+		if(child.Size() <= 1 || child.Token(1).empty())
+		{
+			child.PrintTrace("missing path");
+			return false;
+		}
+		
+		vector<string> paths;
+		paths.emplace_back(child.Token(1));
+		paths.emplace_back(Files::Resources() + child.Token(1));
+		for(const string &source : GameData::Sources())
+			paths.emplace_back(source + child.Token(1));
+		while(!paths.empty() && !Files::Exists(paths.back()))
+			paths.pop_back();
+		if(paths.empty())
+		{
+			child.PrintTrace("path not found");
+			return false;
+		}
+		
+		size_t count = sources.size();
+		if(key == "atlas")
+		{
+			auto glyphs = make_shared<AtlasGlyphs>();
+			if(glyphs->Load(paths.back()))
+				sources.emplace_back(glyphs);
+		}
+		else if(key == "freetype")
+		{
+			auto glyphs = make_shared<FreeTypeGlyphs>();
+			if(glyphs->Load(paths.back(), size))
+				sources.emplace_back(glyphs);
+		}
+		if(count == sources.size())
+		{
+			child.PrintTrace("load failed");
+			return false;
+		}
+	}
+	if(sources.size() == 0)
+	{
+		node.PrintTrace("missing glyph sources (atlas or freetype)");
+		return false;
+	}
+	
+	// Ignore other children.
```

Comment:
?
(You've already stepped through all the children nodes at this point.)

Reply:
I stepped through all supported children. I'm not erroring on unrecognized children so that old code (stable version) can use data of future versions (unstable).

Example: a new type of glyph source is added to Font and a plugin that overrides the font data uses the new feature. The plugin can still provide a fallback source so it's still usable in the stable version.